### PR TITLE
Adapt passes and pipelines to work generically on FunctionLike.

### PR DESCRIPTION
**Pull request by stellaraccident**
* Specifically, enables general purpose use of util.initializer and friends.
* Depends on https://reviews.llvm.org/D115645

### CODE REVIEW

File: iree/compiler/Dialect/VM/IR/VMOps.td [hunk 1]
```diff
@@ -3549,7 +3549,7 @@ def VM_ReturnOp : VM_Op<"return", [
 
 def VM_FailOp : VM_Op<"fail", [
     DeclareOpInterfaceMethods<VM_SerializableOpInterface>,
-    HasParent<"IREE::VM::FuncOp">,
+    //HasParent<"IREE::VM::FuncOp">,
```
Comment (benvanik) [inline iree/compiler/Dialect/VM/IR/VMOps.td:3552]: can just remove this (I think it's fine for fails to live anywhere)

File: iree/compiler/Dialect/HAL/Transforms/ResolveEntryPointOrdinals.cpp [hunk 2]
```diff
@@ -85,6 +87,7 @@ class ResolveEntryPointOrdinalsPass
   void runOnOperation() override {
     MLIRContext *context = &getContext();
     OwningRewritePatternList patterns(&getContext());
+    getOperation().dump();
```
Comment (benvanik) [inline iree/compiler/Dialect/HAL/Transforms/ResolveEntryPointOrdinals.cpp:90]: extra

**Review by benvanik — APPROVED** [review]

File: iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp [hunk 3]
```diff
@@ -149,13 +151,24 @@ class OutlineDispatchRegionsPass
...
+    for (auto it : llvm::enumerate(getOperation().getOps())) {
+      Operation &op = it.value();
+      if (!op.hasTrait<OpTrait::FunctionLike>()) continue;
+
+      // Generate a nice name if possible.
+      std::string opName;
+      if (auto funcOp = llvm::dyn_cast<mlir::FuncOp>(op)) {
+        opName = funcOp.getName().str();
+      } else {
+        opName = std::string("_function_like_") + std::to_string(it.index());
```
Comment (benvanik) [inline iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp:163]: would be nice to have a case for initializers in here, given that they're the only other one we have now and seeing `_function_like_` everywhere would make me sad (`_initializer_` is probably good enough).

**Review by benvanik — APPROVED** [review]

File: iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp (hunk 3 above)
Comment (stellaraccident) [reply iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp:163]: Sold.

**Pull request merged**